pkg/controller/elasticsearch/pdb/roles.go

```diff
@@ -0,0 +1,518 @@
+	// For simplicity both single and multi-statefulsets use matchExpressions with In operator
+	return &metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      label.ClusterNameLabelName,
+				Operator: metav1.LabelSelectorOpIn,
+				Values:   []string{es.Name},
+			},
+			{
+				Key:      label.StatefulSetNameLabelName,
+				Operator: metav1.LabelSelectorOpIn,
+				Values:   ssetNames,
+			},
+		},
+	}
+}
+
+// reconcileAndDeleteUnnecessaryPDBs reconciles the PDBs that are expected to exist and deletes any that exist but are not expected.
+func reconcileAndDeleteUnnecessaryPDBs(ctx context.Context, k8sClient k8s.Client, es esv1.Elasticsearch, expectedPDBs []*policyv1.PodDisruptionBudget) error {
+	existingPDBs, err := listAllRoleSpecificPDBs(ctx, k8sClient, es)
+	if err != nil {
+		return fmt.Errorf("while listing existing role-specific PDBs: %w", err)
+	}
+
+	toDelete := make(map[string]policyv1.PodDisruptionBudget)
+
+	// Populate the toDelete map with existing PDBs
+	for _, pdb := range existingPDBs {
+		toDelete[pdb.Name] = pdb
+	}
+
+	// Remove expected PDBs from the toDelete map
+	for _, pdb := range expectedPDBs {
+		delete(toDelete, pdb.Name)
+		// Ensure that the expected PDB is reconciled.
+		if err := reconcilePDB(ctx, k8sClient, es, pdb); err != nil {
+			return fmt.Errorf("while reconciling role-specific PDB %s: %w", pdb.Name, err)
+		}
+	}
+
+	// Delete unnecessary PDBs
+	for name, pdb := range toDelete {
+		if err := k8sClient.Delete(ctx, &pdb); err != nil {
+			return fmt.Errorf("while deleting role-specific PDB %s: %w", name, err)
+		}
+	}
+
+	return nil
+}
+
+// listAllRoleSpecificPDBs lists all role-specific PDBs for the cluster by retrieving
+// all PDBs in the namespace with the cluster label and verifying the owner reference.
+func listAllRoleSpecificPDBs(ctx context.Context, k8sClient k8s.Client, es esv1.Elasticsearch) ([]policyv1.PodDisruptionBudget, error) {
+	// List all PDBs in the namespace with the cluster label
+	var pdbList policyv1.PodDisruptionBudgetList
+	if err := k8sClient.List(ctx, &pdbList, client.InNamespace(es.Namespace), client.MatchingLabels{
+		label.ClusterNameLabelName: es.Name,
+	}); err != nil {
+		return nil, err
+	}
+
+	// Filter only PDBs that are owned by this Elasticsearch controller
+	var roleSpecificPDBs []policyv1.PodDisruptionBudget
+	for _, pdb := range pdbList.Items {
+		// Check if this PDB is owned by the Elasticsearch resource
+		if isOwnedByElasticsearch(pdb, es) {
+			roleSpecificPDBs = append(roleSpecificPDBs, pdb)
+		}
+	}
+	return roleSpecificPDBs, nil
+}
+
+// deleteAllRoleSpecificPDBs deletes all existing role-specific PDBs for the cluster by retrieving
+// all PDBs in the namespace with the cluster label and verifying the owner reference.
+func deleteAllRoleSpecificPDBs(ctx context.Context, k8sClient k8s.Client, es esv1.Elasticsearch) error {
```

Comment:
```suggestion
// deleteAllPDBs deletes all existing PDBs for the cluster by retrieving
// all PDBs in the namespace with the cluster label and verifying the owner reference.
func deleteAllPDBs(ctx context.Context, k8sClient k8s.Client, es esv1.Elasticsearch) error {
```

Or maybe I'm missing how we are targeting "role specific" PDBs.

Reply:
Ok, I'm handling both v1 and v1alpha1 logic now, but it has made the code a bit ugly. Lmk if you see optimizations.